### PR TITLE
SortOptions on load

### PR DESCRIPTION
**Pull request by oshi97**
This commit adds the ability for SortOptions to be applied
on page load and back navigation, using the new sortBys url param.
Now, the sortBys state is stored in the URL. The SortOptions component
will update it's component state if it sees a persisted sortBys
option that exists in its config.

J=SLAP-1090
TEST=manual, auto

test that I can apply sort options on load, and back nav
see that it works when back/forward naving to either a blank url,
or a url with sortBys in it

### CODE REVIEW

File: src/core/storage/storagekeys.js [hunk 1]
```diff
@@ -32,7 +32,7 @@ const StorageKeys = {
   SESSIONS_OPT_IN: 'sessions-opt-in',
   VERTICAL_PAGES_CONFIG: 'vertical-pages-config',
   LOCALE: 'locale',
-  SORT_BYS: 'sort-bys',
+  SORT_BYS: 'sortBys',
```
Comment (cea2aj) [inline src/core/storage/storagekeys.js:36]: for my own edification, what is the reason for this change?

Comment (oshi97) [reply src/core/storage/storagekeys.js:36]: In the spec we wanted to have url params that mirror liveApis for setting stuff like sorts/filters on load. The api param is `sortBys` not `sort-bys`, so it seemed like the easiest way to do this was to 
1. rename it to sortBys, so that when it shows up in the urls it's `sortBys` not `sort-bys`, and 
2. do setWithPersists on `SORT_BYS`, instead of having a separate `PERSISTED_SORT_BYS` or something. 

The other PRs had to have new storage keys, because some transformations take place between the data in storage and the data sent to the api, but for `SORT_BYS` what we store in storage is exactly what is sent to the api, so it seemed like we could get away with this instead?

File: tests/ui/components/filters/sortoptionscomponent.js [hunk 2]
```diff
@@ -182,4 +183,60 @@ describe('sort options component', () => {
...
+  it('sets persisted sort option on back navigation', () => {
+    const opts = {
+      ...defaultConfig,
+      options: threeOptions
+    };
+    const component = COMPONENT_MANAGER.create('SortOptions', opts);
+    const wrapper = mount(component);
+    expect(wrapper.find('.yxt-SortOptions-optionSelector').at(0).getDOMNode().checked).toBeTruthy();
+    COMPONENT_MANAGER.core.storage.setWithPersist(StorageKeys.SORT_BYS, [threeOptions[2]]);
+    COMPONENT_MANAGER.core.storage.set(StorageKeys.HISTORY_POP_STATE, {});
```
Comment (cea2aj) [inline tests/ui/components/filters/sortoptionscomponent.js:224]: Can you expand on how this properly simulates a back nav? Is updating `HISTORY_POP_STATE` sufficient to fully simulate it? Would an acceptance test be a better fit for this? For a back nav test, I would expect the test to check the option on one page, change the page, and then hit the back button and check again

File: src/ui/components/filters/sortoptionscomponent.js [hunk 3]
```diff
@@ -43,6 +43,33 @@ export default class SortOptionsComponent extends Component {
...
+      }
+    });
+  }
+
+  /**
+   * Returns the option index matching the persisted sortBys, if one exists.
+   *
+   * @returns {number|undefined}
+   */
+  getPersistedOptionIndex () {
```
Comment (cea2aj) [inline src/ui/components/filters/sortoptionscomponent.js]: To clarify, this function is returning the selected persisted option, correct? If so, what do you think of renaming this to `getSelectedPersistedOptionIndex`?

File: src/ui/components/filters/sortoptionscomponent.js [hunk 4]
```diff
@@ -43,6 +43,33 @@ export default class SortOptionsComponent extends Component {
...
+    const persistedSortBys = this.core.storage.get(StorageKeys.SORT_BYS) || [];
```
Comment (cea2aj) [inline src/ui/components/filters/sortoptionscomponent.js:64]: Why is persistedSortBys an array? Aren't we just persisting one  (the selected one)?

File: src/core/storage/storagekeys.js (hunk 1 above)
Comment (cea2aj) [reply src/core/storage/storagekeys.js:36]: ah, okay. that makes sense

File: src/ui/components/filters/sortoptionscomponent.js (hunk 3 above)
Comment (oshi97) [reply src/ui/components/filters/sortoptionscomponent.js]: Technically the option isn't selected yet though. It seems a little weird to call it that when the method has nothing to do with whether the option is selected or not

File: src/ui/components/filters/sortoptionscomponent.js (hunk 4 above)
Comment (oshi97) [reply src/ui/components/filters/sortoptionscomponent.js:64]: The api parameter takes in an array, it's just that the SDK is only set up to let you sort by one "sort by" at a time

File: tests/ui/components/filters/sortoptionscomponent.js (hunk 2 above)
Comment (oshi97) [reply tests/ui/components/filters/sortoptionscomponent.js:224]: Sure, basically the SortOptions component updates itself on backnav by listening for a HISTORY_POP_STATE event. It doesn't do anything else on back nav. SortOptions is a lot simpler than Facets/FilterBox/FilterOptions, and doesn't interact with other components in weird ways, so I thought a regular jest test would be enough here.

File: src/ui/components/filters/sortoptionscomponent.js (hunk 4 above)
Comment (oshi97) [reply src/ui/components/filters/sortoptionscomponent.js:64]: The SDK also internally stores an array for SORT_BYS, though it'll only ever have one element, since SortOptions only ever calls core.setSortBys() with one element. (I guess you could directly call core.setSortBys with multiple elements in your config/onReady)

File: src/ui/components/filters/sortoptionscomponent.js (hunk 3 above)
Comment (cea2aj) [reply src/ui/components/filters/sortoptionscomponent.js]: ah, I see. But we do only persist the selected option, right?

File: tests/ui/components/filters/sortoptionscomponent.js [hunk 5]
```diff
@@ -182,4 +183,60 @@ describe('sort options component', () => {
...
+    wrapper.update();
+    expect(wrapper.find('.yxt-SortOptions-optionSelector').at(3).getDOMNode().checked).toBeTruthy();
```
Comment (cea2aj) [inline tests/ui/components/filters/sortoptionscomponent.js:226]: Why are we checking at index 3 if there are only three options in the array?

File: tests/ui/components/filters/sortoptionscomponent.js [hunk 6]
```diff
@@ -182,4 +183,60 @@ describe('sort options component', () => {
...
+
+  it('sets persisted sort option on back navigation', () => {
+    const opts = {
+      ...defaultConfig,
+      options: threeOptions
+    };
+    const component = COMPONENT_MANAGER.create('SortOptions', opts);
+    const wrapper = mount(component);
+    expect(wrapper.find('.yxt-SortOptions-optionSelector').at(0).getDOMNode().checked).toBeTruthy();
+    COMPONENT_MANAGER.core.storage.setWithPersist(StorageKeys.SORT_BYS, [threeOptions[2]]);
```
Comment (cea2aj) [inline tests/ui/components/filters/sortoptionscomponent.js:223]: Should we also be pushing the state to history?

File: tests/ui/components/filters/sortoptionscomponent.js (hunk 2 above)
Comment (cea2aj) [reply tests/ui/components/filters/sortoptionscomponent.js:224]: okay, that makes sense. I guess I'm still trying to figure out why this test doesn't have three expects: One for the initial page, a second after navigating away to another page, and a third after hitting the back button. And shouldn't the first and third expect have the same option selected?

File: src/ui/components/filters/sortoptionscomponent.js (hunk 3 above)
Comment (oshi97) [reply src/ui/components/filters/sortoptionscomponent.js]: Ah yeah that's true. What do you think about `getPersistedSelectedOptionIndex` then?

File: tests/ui/components/filters/sortoptionscomponent.js (hunk 5 above)
Comment (oshi97) [reply tests/ui/components/filters/sortoptionscomponent.js:226]: sort options adds a default option even if you have no options (very weird I know, I think this was part of the original SortOptions spec)

File: tests/ui/components/filters/sortoptionscomponent.js (hunk 6 above)
Comment (oshi97) [reply tests/ui/components/filters/sortoptionscomponent.js:223]: no, SortOptions doesn't care about the actual history state, just the HISTORY_POP_STATE event. (Also afaik jsdom still doesn't support navigation)

File: src/ui/components/filters/sortoptionscomponent.js (hunk 3 above)
Comment (cea2aj) [reply src/ui/components/filters/sortoptionscomponent.js]: that works!

File: tests/ui/components/filters/sortoptionscomponent.js (hunk 2 above)
Comment (cea2aj) [reply tests/ui/components/filters/sortoptionscomponent.js:224]: Actually, I see now. All we really have to do in this test is to ensure that it responds to the history pop state change. What do you think about renaming the test to something like "uses the persisted sorts when the history pop state updates"?

Comment (oshi97) [reply tests/ui/components/filters/sortoptionscomponent.js:224]: This test is just for back navigating from a page. I guess to be a little more accurate, this isn't really for "back navigation" so much as "when HISTORY_POP_STATE is updated, reset the component state using persisted storage". I can add another case that tests going back and forth

File: src/ui/components/filters/sortoptionscomponent.js [hunk 7]
```diff
@@ -190,6 +216,7 @@ export default class SortOptionsComponent extends Component {
    * Trigger a search with all filters in storage
    */
   _search () {
+    console.log('trigger sort search');
```
Comment (tmeyer2115) [inline src/ui/components/filters/sortoptionscomponent.js]: Should this be removed?

Comment (oshi97) [reply src/ui/components/filters/sortoptionscomponent.js]: 🤦  yes, updated!

**Review by tmeyer2115 — APPROVED** [review]
